### ISSUE TITLE
Bugs/issues after the big destructuring PR

### ISSUE THREAD

**Issue by MStoykov** [dop251/goja#305]
First great job on implementing all of that functionality :clap: :bow: :tada: and thank you very much.

As I have previously mentioned k6 runs the majority (no async/await, promise stuff) of the tc39 testsuite including tests that fail (for various reasons) to test the goja+babel setup we have and its problems. As such we have some *new* tests that are now failing without disabling plugins (and even more that are now passing :clap: , unfortunately, I am here to only tell you about the failing ones, sorry :(  ):

- [language/expressions/assignmenttargettype/parenthesized-primaryexpression-objectliteral.js ](https://github.com/tc39/test262/blob/0efd2a3baa4e2b00a1328bbf218ac835943ccbe4/test/language/expressions/assignmenttargettype/parenthesized-primaryexpression-objectliteral.js) - `({}) = 1;` should be a syntax error
- [language/expressions/call/scope-lex-open.js](https://github.com/tc39/test262/blob/0efd2a3baa4e2b00a1328bbf218ac835943ccbe4/test/language/expressions/call/scope-lex-open.js) - this now fails with `Duplicate declaration n` 
- [language/expressions/object/getter-param-dflt.js](https://github.com/tc39/test262/blob/0efd2a3baa4e2b00a1328bbf218ac835943ccbe4/test/language/expressions/object/getter-param-dflt.js) - something about methods :man_shrugging: 
- [language/statements/for-of/scope-head-lex-close.js](https://github.com/tc39/test262/blob/d19534c3ce2fa23d2a079059f23b36126b2e66d4/test/language/statements/for-of/scope-head-lex-close.js) panics with `runtime error: invalid memory address or nil pointer dereference` as well as it's ...lex-open.js test and the same two tests for `for-in` 

Then of course I started disabling babel plugins of now implemented stuff ;).
Disabling the `transform-es2015-destructuring` and `transform-es2015-computed-properties` plugins of babel also shows that:
- a lot of tests for `for-of` destructuring break as in [this one](https://github.com/tc39/test262/blob/d19534c3ce2fa23d2a079059f23b36126b2e66d4/test/language/statements/for-of/dstr/array-elem-iter-nrml-close-skip.js) with  "Line 55:1 Invalid left-hand side in for-in or for-of" I would expect that ... most of those are the same issue so I am not listing all of them.
- same for `try` in [this test](https://github.com/tc39/test262/blob/d19534c3ce2fa23d2a079059f23b36126b2e66d4/test/language/statements/try/dstr/ary-ptrn-elem-ary-rest-init.js) with "Line 36:10 Unexpected token ["  which seems more like a case of destructuring not being support in the catch parameter.

Disabling `transform-es2015-parameters` breaks *a lot* of tests and I didn't want to go into it as I was pretty sure I did something wrong on my side, but it seems to not be me :cry: :sob: (edit: opened https://github.com/dop251/goja/issues/309). The same goes for `transform-es2015-spread` which also makes the tests take forever, probably because of all the failures. On the other hand ... it really shouldn't so I will keep digging in those two babel plugins to see what is going on and report back when I figure what is going on.

**Comment by dop251** [comment]
Thanks for reporting these. I think I have fixed most of them, except for the patterns in catch, shall have a look later.

**Comment by MStoykov** [comment]
Thanks very much for the extremely fast fixes :bow: 

I did rerun my tests and apart from the try/catch stuff all seems to now work.

Additionally, I now noticed that tests such as [language/eval-code/direct/func-decl-a-following-parameter-is-named-arguments-declare-arguments-and-assign.js](https://github.com/tc39/test262/blob/main/test/language/eval-code/direct/func-decl-a-following-parameter-is-named-arguments-declare-arguments-and-assign.js) are now failing due to not throwing SyntaxError where it's expected instead of the previous error ("arguments is a reserved word in strict mode") that shows me, that we(k6) probably shouldn't run this test in k6 as it's `noStrict` and we always run in `strict` mode in k6  ... so yeah. But it still doesn't work in goja from my quick test as well.
 
While I would think this particular (or all the other tests failing around this) aren't really practical in themself this seems to be some kind of evaluation ordering bug(?) :thinking: so it seems interesting from that perspective.
